File: scripts/fix_unused_vars.js
const { spawnSync } = require('child_process');
const fs = require('fs');
const path = require('path');

const root = process.cwd();

function runEslint() {
  const result = spawnSync('C:\\Windows\\System32\\cmd.exe', ['/d', '/s', '/c', 'npx eslint src/app --ext .ts,.tsx --format json'], {
    cwd: root,
    encoding: 'utf8',
  });

  const stdout = result.stdout || '';
  return JSON.parse(stdout || '[]');
}

function removeImportUnused(line, unusedNames) {
  const importMatch = line.match(/^(import\s+(?:type\s+)?(?:[^\s,]+\s*,\s*)?(?:\{(?<specs>[^}]*)\}|(?<default>[^\s,]+))(?<rest>.*?))$/);
  if (!importMatch) return null;

  const specs = importMatch.groups.specs;
  const defaultName = importMatch.groups.default;
  if (specs !== undefined) {
    const parts = specs.split(',').map((p) => p.trim()).filter(Boolean);
    const keep = parts.filter((p) => !unusedNames.has(p.split(' as ').pop().trim()));
    if (keep.length) {
      return line.replace(specs, keep.join(', '));
    }
    return null;
  }

  if (defaultName && unusedNames.has(defaultName)) {
    return null;
  }
  return line;
}

function updateDestructuredSignature(line, unusedNames) {
  const sig = line.match(/function\s+\w+\s*\(([^)]*)\)/);
  if (!sig) return line;

  const params = sig[1];
  const destructMatch = params.match(/\(\{\s*([^}]*)\s*\}\s*:\s*(\{[^)]*\})\)/);
  if (!destructMatch) return line;

  const names = destructMatch[1].split(',').map((n) => n.trim()).filter(Boolean).map((n) => n.split('=')[0].trim().split(' as ').pop().trim());
  const keep = names.filter((name) => !unusedNames.has(name));
  if (keep.length === names.length) return line;

  const replacement = keep.length ? keep.join(', ') : '';
  let newParams = params;
  if (keep.length) {
    newParams = newParams.replace(destructMatch[1], replacement);
  } else {
    const typeAnnotation = destructMatch[2];
    newParams = newParams.replace(destructMatch[0], '({}: ' + typeAnnotation + ')');
  }
  return line.replace(params, newParams);
}

function updateAssignmentDeclaration(line, unusedNames) {
  const arrMatch = line.match(/^(\s*(?:const|let|var)\s+)\[(.*)\]\s*=.*$/);
  if (arrMatch) {
    const parts = arrMatch[2].split(',').map((p) => p.trim()).filter(Boolean).map((p) => p.split('=')[0].trim());
    const keep = parts.filter((name) => !unusedNames.has(name));
    if (keep.length !== parts.length) {
      if (keep.length) {
        const rest = line.split('=')[1];
        return `${arrMatch[1]}[${keep.join(', ')}] =${rest}`;
      }
      return null;
    }
  }

  const objMatch = line.match(/^(\s*(?:const|let|var)\s+)\{(.*)\}\s*=.*$/);
  if (objMatch) {
    const parts = objMatch[2].split(',').map((p) => p.trim()).filter(Boolean).map((p) => p.split('=')[0].trim());
    const keep = parts.filter((name) => !unusedNames.has(name));
    if (keep.length !== parts.length) {
      if (keep.length) {
        const rest = line.split('=')[1];
        return `${objMatch[1]}{${keep.join(', ')}} =${rest}`;
      }
      return null;
    }
  }

  const simpleMatch = line.match(/^(\s*(?:const|let|var)\s+)([A-Za-z_][A-Za-z0-9_]*)(\s*=.*)$/);
  if (simpleMatch && unusedNames.has(simpleMatch[2])) return null;

  return line;
}

function main() {
  const items = runEslint();

  for (const entry of items) {
    const filePath = path.resolve(root, entry.filePath.replace(/^C:\\projects\\nextprojects\\higorgeous\\/, ''));
    if (!fs.existsSync(filePath)) continue;

    const lines = fs.readFileSync(filePath, 'utf8').split(/\r?\n/);
    const issuesByLine = {};

    for (const msg of entry.messages || []) {
      if (!['@typescript-eslint/no-unused-vars', 'no-unused-vars'].includes(msg.ruleId)) continue;
      const nameMatch = msg.message.match(/'([^']+)'/);
      if (!nameMatch) continue;
      const name = nameMatch[1];
      if (!issuesByLine[msg.line - 1]) issuesByLine[msg.line - 1] = new Set();
      issuesByLine[msg.line - 1].add(name);
    }

    let changed = false;
    for (const lineNo of Object.keys(issuesByLine).sort((a, b) => Number(b) - Number(a))) {
      const unusedNames = issuesByLine[lineNo];
      let line = lines[Number(lineNo)];
      let stripped = line.trim();

      if (stripped.startsWith('import ')) {
        const newLine = removeImportUnused(line, unusedNames);
        if (newLine === null) {
          lines[Number(lineNo)] = '';
          changed = true;
          continue;
        }
        if (newLine !== line) {
          lines[Number(lineNo)] = newLine;
          changed = true;
          continue;
        }
      }

      const sigUpdated = updateDestructuredSignature(line, unusedNames);
      if (sigUpdated !== line) {
        lines[Number(lineNo)] = sigUpdated;
        changed = true;
        continue;
      }

      const declUpdated = updateAssignmentDeclaration(line, unusedNames);
      if (declUpdated === null) {
        lines[Number(lineNo)] = '';
        changed = true;
        continue;
      }
      if (declUpdated !== line) {
        lines[Number(lineNo)] = declUpdated;
        changed = true;
      }
    }

    if (changed) {
      const cleaned = lines.filter((ln) => ln.trim() !== '').join('\n').trimEnd() + '\n';
      fs.writeFileSync(filePath, cleaned, 'utf8');
    }
  }
}

main();

File: scripts/fix_unused_vars.py
import json
import re
import subprocess
from pathlib import Path


ROOT = Path('c:/projects/nextprojects/higorgeous').resolve()


def run_eslint():
    result = subprocess.run(
        ['npx', 'eslint', 'src/app', '--ext', '.ts,.tsx', '--format', 'json'],
        capture_output=True,
        text=True,
        cwd=ROOT,
        shell=False,
    )
    return json.loads(result.stdout) if result.stdout.strip() else []


def remove_import_unused(line: str, unused_names: set[str]):
    # Handles named imports such as: import { A, B } from 'x';
    import_match = re.match(r'^(import\s+(?:type\s+)?(?:[^\s,]+\s*,\s*)?(?:\{(?P<specs>[^}]*)\}|(?P<default>[^\s,]+))(?P<rest>.*?))$', line)
    if not import_match:
        return None

    specs = import_match.group('specs')
    default = import_match.group('default')
    if specs is not None:
        parts = [p.strip() for p in specs.split(',') if p.strip()]
        keep = [p for p in parts if p.split(' as ')[-1].strip() not in unused_names]
        if keep:
            return line.replace(specs, ', '.join(keep))
        return None

    if default is not None and default in unused_names:
        return None
    return line


def update_destructured_signature(line: str, unused_names: set[str]):
    sig = re.search(r'function\s+\w+\s*\(([^)]*)\)', line)
    if not sig:
        return line
    params = sig.group(1)
    destruct_match = re.search(r'\(\{\s*([^}]*)\s*\}\s*:\s*(\{[^)]*\})\)', params)
    if not destruct_match:
        return line

    names = [n.strip().split('=')[0].strip().split(' as ')[-1].strip()
             for n in destruct_match.group(1).split(',') if n.strip()]
    keep = [n for n in names if n not in unused_names]
    if len(keep) == len(names):
        return line

    replacement = ', '.join(keep) if keep else ''
    new_params = params
    if keep:
        new_params = new_params.replace(destruct_match.group(1), replacement)
    else:
        # replace the whole destructured parameter with an empty object
        new_params = new_params.replace(destruct_match.group(0), '({}: ' + destruct_match.group(2) + ')')
    return line.replace(params, new_params)


def update_assignment_declaration(line: str, unused_names: set[str]):
    # const [a, b] = useState(...)
    arr_match = re.match(r'^(\s*(?:const|let|var)\s+)\[(.*)\]\s*=.*$', line)
    if arr_match:
        parts = [p.strip().split('=')[0].strip() for p in arr_match.group(2).split(',') if p.strip()]
        keep = [p for p in parts if p not in unused_names]
        if len(keep) != len(parts):
            if keep:
                rest = line.split('=', 1)[1]
                return f"{arr_match.group(1)}[{', '.join(keep)}] = {rest.lstrip()}"
            return None

    # const { a, b } = ...
    obj_match = re.match(r'^(\s*(?:const|let|var)\s+)\{(.*)\}\s*=.*$', line)
    if obj_match:
        parts = [p.strip().split('=')[0].strip() for p in obj_match.group(2).split(',') if p.strip()]
        keep = [p for p in parts if p not in unused_names]
        if len(keep) != len(parts):
            if keep:
                rest = line.split('=', 1)[1]
                return f"{obj_match.group(1)}{{{', '.join(keep)}}} = {rest.lstrip()}"
            return None

    # const foo = ...
    simple_match = re.match(r'^(\s*(?:const|let|var)\s+)([A-Za-z_][A-Za-z0-9_]*)(\s*=.*)$', line)
    if simple_match and simple_match.group(2) in unused_names:
        return None

    return line


def main():
    items = run_eslint()

    for entry in items:
        file_path = Path(entry['filePath'])
        if not file_path.exists():
            continue

        lines = file_path.read_text(encoding='utf-8').splitlines()
        issues_by_line = {}
        for msg in entry.get('messages', []):
            if msg.get('ruleId') not in ('@typescript-eslint/no-unused-vars', 'no-unused-vars'):
                continue
            name_match = re.search(r"'([^']+)'", msg.get('message', ''))
            if not name_match:
                continue
            issues_by_line.setdefault(msg['line'] - 1, set()).add(name_match.group(1))

        changed = False
        for line_no in sorted(issues_by_line.keys(), reverse=True):
            unused_names = issues_by_line[line_no]
            line = lines[line_no]
            stripped = line.strip()

            if stripped.startswith('import '):
                new_line = remove_import_unused(line, unused_names)
                if new_line is None:
                    lines[line_no] = ''
                    changed = True
                    continue
                if new_line != line:
                    lines[line_no] = new_line
                    changed = True
                    continue

            new_line = update_destructured_signature(line, unused_names)
            if new_line != line:
                lines[line_no] = new_line
                changed = True
                continue

            new_line = update_assignment_declaration(line, unused_names)
            if new_line is None:
                lines[line_no] = ''
                changed = True
                continue
            if new_line != line:
                lines[line_no] = new_line
                changed = True

        if changed:
            cleaned = '\n'.join([ln for ln in lines if ln.strip() != '']).rstrip() + '\n'
            file_path.write_text(cleaned, encoding='utf-8')


if __name__ == '__main__':
    main()
